src/node_io.h

```diff
@@ -0,0 +1,236 @@
+#ifndef SRC_NODE_IO_H_
+#define SRC_NODE_IO_H_
+
+#include "node.h"
+
+namespace node {
+
+namespace io {
+
+enum io_error {
+  // Source would overflow the provided buffer
+  IO_ERROR_OVERFLOW = -256,
+  // Source does not support FLAG_PEEK
+  IO_ERROR_NO_PEEK = -512,
+  // Attempt to pull from an already ended source
+  IO_ERROR_EOF = -768,
+  // Source will not copy data into caller provided buffer
+  IO_ERROR_MUST_NO_COPY = -1024,
+};
+
+enum io_pull_flags {
+  IO_PULL_FLAG_NONE = 0x0,
+  // Callback must be invoked synchronously (also sets the
+  // IO_PULL_FLAG_MUST_COPY and IO_PULL_FLAG_STRICT_LENGTH flags)
+  IO_PULL_FLAG_SYNC = 0xD,
+  // Peek at what data is available, do not actually read.
+  IO_PULL_FLAG_PEEK = 0x2,
+  // Puller considers the length given to be strict,
+  // Source must not overflow
```

Comment:
overflow? i.e. return more data than requested?

Reply:
Yes. There are a couple of options here:

1. The caller allocates it's own `io_buf_t` with a specific size, tells the `Source` that it must fill those (this implies that overflowing is not allowed)

2. The caller allocates it's own `io_buf_t` with a specific size, tells the `Source` to provide data up to `length` and no more. `Source` may choose to fill the provided `Buffer`, or provide pointers to it's own allocated buffers. The Sink may or may not allow the `Source` provided buffers to be larger than what it asked for.